File: scala-part/src/main/scala/ru/nstu/scala/part/ui/ListActionListenerImpl.scala
package ru.nstu.scala.part.ui

import ru.nstu.java.part.data.{IList, List, ListUtils}
import ru.nstu.java.part.ui.AbstractListActionListener
import ru.nstu.scala.part.data.MyList

import java.io.FileNotFoundException

class ListActionListenerImpl extends AbstractListActionListener {

  var items: IList[Object] = new MyList[Object]

  override def onAdd(text: String): Unit = {
    if (text.isEmpty) return
    val value = builder.createFromString(text)
    items.add(value)
    listModel.addElement(value)
  }

  override def onInsert(text: String, index: Int): Unit = {
    if (text.isEmpty) return
    val value = builder.createFromString(text)
    items.add(value, index)
    listModel.add(index, value)
  }

  override def onRemove(index: Int): Unit = {
    items.remove(index)
    listModel.remove(index)
  }

  override def onSort(): Unit = {
    items.sort(builder.getComparator)
    listModel.clear()
    items.forEach(el => listModel.addElement(el))
  }

  override def onSave(): Unit = {

    try ListUtils.saveToFile(filename, items, builder)
    catch {
      case e: FileNotFoundException =>
        System.err.println("Unable to write list to a file")
        e.printStackTrace()
    }
  }

  override def onLoad(): Unit = {
    try {
      items = ListUtils.loadFromFile(filename, builder, new List[AnyRef])
      listModel.clear()
      items.forEach(el => listModel.addElement(el))
    } catch {
      case e: Exception =>
        System.err.println("Unable to read list from a file")
        e.printStackTrace()
    }
  }
}

File: java-part/src/main/java/ru/nstu/java/part/ui/ListActionListenerImpl.java
package ru.nstu.java.part.ui;

import ru.nstu.java.part.data.IList;
import ru.nstu.java.part.data.List;
import ru.nstu.java.part.data.ListUtils;

import java.io.FileNotFoundException;

public class ListActionListenerImpl extends AbstractListActionListener {


    protected IList<Object> items = new List<>();

    @Override
    public void onAdd(String text) {
        if (text.isEmpty()) return;
        Object data = builder.createFromString(text);
        items.add(data);
        listModel.addElement(data);
    }

    @Override
    public void onInsert(String text, int index) {
        if (text.isEmpty()) return;
        Object data = builder.createFromString(text);
        items.add(data, index);
        listModel.add(index, data);
    }

    @Override
    public void onRemove(int index) {
        items.remove(index);
        listModel.remove(index);
    }

    @Override
    public void onSort() {
        items.sort(builder.getComparator());
        listModel.clear();
        items.forEach(listModel::addElement);
    }

    @Override
    public void onSave() {
        try {
            ListUtils.saveToFile(filename, items, builder);
        } catch (FileNotFoundException e) {
            System.err.println("Unable to write list to a file");
            e.printStackTrace();
        }
    }

    @Override
    public void onLoad() {
        try {
            items = ListUtils.loadFromFile(filename, builder, new List<>());
            listModel.clear();
            items.forEach(listModel::addElement);
        } catch (Exception e) {
            System.err.println("Unable to read list from a file");
            e.printStackTrace();
        }
    }
}
